### PR TITLE
Fix signal related contentions and bugs

### PR DESCRIPTION
**Pull request by krnowak**
This PR is to fix issue #8.

First commit adds a bunch of tests for testing contentions. Also, while working on this issue I noticed that Signal function does not behave as it should, so I added tests for it as well.

Second commit should fix the contentions.

Third commit should fix the signal watchers adding and removing, so all the tests I've added in first commit pass.

Fourth commit is there, because "appropriate" is hard to type. :)

Fifth commit fixes some confusing docs.

### CODE REVIEW

File: signal_test.go [hunk 1]
```diff
@@ -0,0 +1,490 @@
+package dbus_test
```
Comment (bcwaldon) [inline signal_test.go]: This should be `package dbus`

File: signal_test.go [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+
+// Creates two separate connections to system bus, one for server, one
+// for client, creates a server and adds two channels for signal
+// watching.
+//
+// We are adding two unbufferred channels to watch signals, so we can
+// watch both of them in select, handle one of them and then be sure
+// that dbus signal handling is stuck on sending signal to other
+// channel.
+func testSetup(test *testing.T) *TestData {
```
Comment (bcwaldon) [inline signal_test.go]: golang convention is to use `t` not `test`

File: signal_test.go [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+		server.test.Log("Server: Emitting belayed", orderReceived, "signal")
+		err := server.conn.Emit(objectPath, orderReceivedFull)
+		if err != nil {
+			panic(err)
+		}
+	}()
+	return nil
+}
+
+// It is like dbus.SessionBus() but it creates new connection
```
Comment (bcwaldon) [inline signal_test.go]: Please format docstrings per https://golang.org/doc/effective_go.html#commentary

File: conn.go [hunk 4]
```diff
@@ -347,6 +376,97 @@ func (conn *Conn) inWorker() {
 	}
 }
 
+// sigWorker runs in an own goroutine, getting signals and handling
```
Comment (bcwaldon) [inline conn.go]: This comment is confusing - it's not inherently running in an independent goroutine. Are you trying to say that it *should* run in its own goroutine?

File: conn.go [hunk 5]
```diff
@@ -347,6 +376,97 @@ func (conn *Conn) inWorker() {
...
+		signal := conn.nextSignal()
+		if signal != nil {
+			closing := conn.handleSignal(signal)
+			if closing {
+				break WorkerLoop
+			}
+		} else {
+			select {
+			case <-conn.signalQueueWakeUp:
+				continue WorkerLoop
```
Comment (bcwaldon) [inline conn.go]: You don't need this statement.

File: conn.go [hunk 6]
```diff
@@ -347,6 +376,97 @@ func (conn *Conn) inWorker() {
...
+			case <-conn.shutdownSigWorker:
+				break WorkerLoop
```
Comment (bcwaldon) [inline conn.go]: Rather than break, why not just return?

File: conn.go [hunk 7]
```diff
@@ -347,6 +376,97 @@ func (conn *Conn) inWorker() {
 	}
 }
 
+// sigWorker runs in an own goroutine, getting signals and handling
+// them appropriately.
+func (conn *Conn) sigWorker() {
+WorkerLoop:
```
Comment (bcwaldon) [inline conn.go]: This label would be unnecessary if you just `return` instead of `break` and drop the `continue` below.

File: conn.go [hunk 8]
```diff
@@ -161,6 +170,8 @@ func newConn(tr transport) (*Conn, error) {
 	conn.nextSerial = 1
 	conn.serialUsed = map[uint32]bool{0: true}
 	conn.busObj = conn.Object("org.freedesktop.DBus", "/org/freedesktop/DBus")
+	conn.shutdownSigWorker = make(chan int, 1)
```
Comment (bcwaldon) [inline conn.go]: When the value doesn't matter, we use an empty struct rather than an int or bool.

File: conn.go [hunk 9]
```diff
@@ -347,6 +376,97 @@ func (conn *Conn) inWorker() {
...
+	}
+
+	conn.doWakeUp = true
+	return nil
+}
+
+// Either adds new pending channels to main signal channels array or
+// removes them if they already existed there.
+//
+// It is assumed that signals lock is locked.
```
Comment (bcwaldon) [inline conn.go]: It would help to be explicit here and refer to the actual variable name rather than just "signals lock"

File: conn.go [hunk 10]
```diff
@@ -347,6 +376,97 @@ func (conn *Conn) inWorker() {
...
+SignalLoop:
+	for _, ch := range conn.signals {
+		select {
+		case ch <- signal:
+		case <-conn.shutdownSigWorker:
+			closing = true
+			break SignalLoop
+		}
+	}
+	conn.signalsLck.Unlock()
```
Comment (bcwaldon) [inline conn.go]: Just defer this after calling Lock above

File: conn.go [hunk 11]
```diff
@@ -347,6 +376,97 @@ func (conn *Conn) inWorker() {
...
+
+	conn.processPendingSignalChannels()
+
+	closing := false
+SignalLoop:
+	for _, ch := range conn.signals {
+		select {
+		case ch <- signal:
+		case <-conn.shutdownSigWorker:
+			closing = true
```
Comment (bcwaldon) [inline conn.go]: Just `return true` here

File: conn.go [hunk 12]
```diff
@@ -347,6 +376,97 @@ func (conn *Conn) inWorker() {
...
+			break SignalLoop
+		}
+	}
+	conn.signalsLck.Unlock()
+	return closing
```
Comment (bcwaldon) [inline conn.go]: Just `return false` here, then the `SignalLoop` label goes away

File: signal_test.go [hunk 13]
```diff
@@ -0,0 +1,490 @@
+package dbus_test
+
+import (
+	"errors"
```
Comment (bcwaldon) [inline signal_test.go]: Break apart imports into three groups:

```
import (
    stdlib packages

    third-party packages

    local packages
)
```

So in this case:

```
import (
    "errors"
    "testing"
    "time"

    "github.com/godbus/dbus/introspect"
)
```

I didn't carry over the `github.com/godbus/dbus` import since this file should actually be a part of that package (see comment above about package naming).

File: conn.go [hunk 14]
```diff
@@ -53,8 +53,17 @@ type Conn struct {
...
+	signalsLck        sync.Mutex
+	shutdownSigWorker chan int
+
+	pendingSignalChannels    []chan<- *Signal
+	pendingSignalChannelsLck sync.Mutex
+
+	signalQueue       []*Signal
+	signalQueueLck    sync.Mutex
+	signalQueueWakeUp chan int
+	doWakeUp          bool
```
Comment (bcwaldon) [inline conn.go]: Should access to this bool always be protected by `signalQueueLck`?

**Comment by bcwaldon** [comment]
@krnowak I haven't looked closely at the possible raciness of this, but I've made enough comments that I'm going to stop here for now.

File: signal_test.go (hunk 2 above)
Comment (krnowak) [inline signal_test.go]: Ok, I just wanted to avoid single-letter variables. Will change it.

File: conn.go (hunk 4 above)
Comment (krnowak) [inline conn.go]: It is running in its own goroutine, in the same way like inWorker and outWorker - see addition in auth.go

File: signal_test.go (hunk 2 above)
Comment (jonboulle) [inline signal_test.go]: I empathise, but they are pretty idiomatic in go for small scopes (or for very well-established patterns like this one)

File: conn.go (hunk 14 above)
Comment (krnowak) [inline conn.go]: It is accessed by two different goroutines (inWorker and sigWorker), so I think it should. And it is.

File: signal_test.go (hunk 1 above)
Comment (krnowak) [inline signal_test.go]: After changing the package and removing github.com/godbus/dbus from imports I'm getting this:

    [CoreOS-dev dbus]$ go test -v
    # github.com/godbus/dbus
    import cycle not allowed in test
    package github.com/godbus/dbus (test)
    	imports github.com/godbus/dbus/introspect
    	imports github.com/godbus/dbus
    
    FAIL	github.com/godbus/dbus [setup failed]

I'm inclined to leave it as is. Maybe with camelCased package name. Or maybe it should be `package introspect` (though the test doesn't have too much to do with this package).

Comment (bcwaldon) [inline signal_test.go]: Folders are intended to map to a single package, so we need to figure out how to make this work. If the tests for the `dbus` package require something in the `introspect` package, that then depends back on the `dbus` package, there are three paths forward:
1. move the tests into the `introspect` package
2. combine the `introspect` and `dbus` packages
3. break the dependency in the `introspect` package on `dbus`

I'd prefer option 2

File: conn.go (hunk 4 above)
Comment (bcwaldon) [inline conn.go]: Being overly explicit here: `sigWorker` does not inherently run in its own goroutine. The only caller of `sigWorker` happens to run it in its own goroutine, but there could be another caller that doesn't do the same. Let's just not document `sigWorker` as running in a goroutine.

File: conn.go (hunk 14 above)
Comment (bcwaldon) [inline conn.go]: Ok - can we document that?

File: signal_test.go (hunk 1 above)
Comment (krnowak) [inline signal_test.go]: Ok, I'll have a look.

File: conn.go (hunk 14 above)
Comment (jonboulle) [inline conn.go]: Convention is that everything the mutex is actually protecting is grouped right below it, right?

File: conn.go (hunk 4 above)
Comment (krnowak) [inline conn.go]: Ok, want me fix inWorker and outWorker docs at the same time? (I based sigWorker's docs on theirs.)

File: conn.go (hunk 14 above)
Comment (krnowak) [inline conn.go]: Yes, will do it.

File: signal_test.go (hunk 1 above)
Comment (krnowak) [inline signal_test.go]: And BTW. That sounds like we don't have to care about stable API?

Comment (krnowak) [inline signal_test.go]: Ok, I sidestepped the issue by removing the use of introspection. It's not mandatory for dbus server and it wasn't used at all in these tests.

File: conn.go (hunk 14 above)
Comment (krnowak) [inline conn.go]: Instead of documenting it, I renamed it to signalQueueConsumerAsleep. Quite a longwinded name, but at least it is clearer that it belongs to the group guarded by signalQueueLck mutex.

File: conn.go [hunk 15]
```diff
@@ -21,6 +21,8 @@ var (
 // ErrClosed is the error returned by calls on a closed connection.
 var ErrClosed = errors.New("dbus: connection closed by user")
 
+type q struct{}
```
Comment (bcwaldon) [inline conn.go]: This feels like an unnecessary layer of indirection to me. Why not just use an empty struct literal?

File: conn.go [hunk 16]
```diff
@@ -322,11 +351,15 @@ func (conn *Conn) inWorker() {
 					Name:   iface + "." + member,
 					Body:   msg.Body,
 				}
-				conn.signalsLck.Lock()
-				for _, ch := range conn.signals {
-					ch <- signal
+				conn.signalQueueLck.Lock()
+				if conn.signalQueueWakeUp != nil {
```
Comment (bcwaldon) [inline conn.go]: When would this ever be nil?

File: conn.go [hunk 17]
```diff
@@ -347,6 +380,93 @@ func (conn *Conn) inWorker() {
...
+	conn.signalQueueConsumerAsleep = true
+	return nil
+}
+
+// processPendingSignalChannels either adds new pending channels to
+// signal channels list or removes them if they already existed there.
+//
+// It is assumed that conn.signalsLck is locked.
+func (conn *Conn) processPendingSignalChannels() {
+	signalsMap := make(map[chan<- *Signal]bool, len(conn.signals))
```
Comment (bcwaldon) [inline conn.go]: Using an empty struct rather than a boolean is more idiomatic here.

File: conn.go [hunk 18]
```diff
@@ -347,6 +380,93 @@ func (conn *Conn) inWorker() {
...
+		}
+	}
+	conn.pendingSignalChannels = nil
+	conn.pendingSignalChannelsLck.Unlock()
+
+	conn.signals = nil
+	for ch, _ := range signalsMap {
+		conn.signals = append(conn.signals, ch)
+	}
+	signalsMap = nil
```
Comment (bcwaldon) [inline conn.go]: Why is this necessary?

File: conn.go [hunk 19]
```diff
@@ -347,6 +380,93 @@ func (conn *Conn) inWorker() {
...
+		} else {
+			signalsMap[ch] = false
+		}
+	}
+	conn.pendingSignalChannels = nil
+	conn.pendingSignalChannelsLck.Unlock()
+
+	conn.signals = nil
+	for ch, _ := range signalsMap {
+		conn.signals = append(conn.signals, ch)
```
Comment (bcwaldon) [inline conn.go]: Are you sure you don't need to copy `ch` before you append it to the slice?

File: conn.go (hunk 15 above)
Comment (krnowak) [inline conn.go]: Less typing. But if that bothers you then I'll get rid of it.

File: conn.go [hunk 20]
```diff
@@ -182,14 +196,29 @@ func (conn *Conn) Close() error {
...
 	}
+
 	close(conn.out)
 	conn.closed = true
 	conn.outLck.Unlock()
+
+	close(conn.shutdownSigWorker)
+	conn.signalQueueLck.Lock()
+	// This is to avoid unnecessary wakeup of sigWorker
+	tmp := conn.signalQueueWakeUp
```
Comment (bcwaldon) [inline conn.go]: I don't actually know how this behaves. Is this copying a reference, or is this actually creating a copy of the channel? I would assume the former, but I can't find any documentation that explains this.

File: conn.go [hunk 21]
```diff
@@ -182,14 +196,29 @@ func (conn *Conn) Close() error {
...
+	conn.signalQueueLck.Unlock()
+
 	conn.signalsLck.Lock()
+	// This is to avoid closing channels user decided to remove
+	// from signal channels list
+	conn.processPendingSignalChannels()
 	for _, ch := range conn.signals {
 		close(ch)
 	}
+	conn.signals = nil
```
Comment (bcwaldon) [inline conn.go]: Why is this necessary?

File: conn.go (hunk 16 above)
Comment (krnowak) [inline conn.go]: I set it to nil when closing the connection to avoid putting another signals in queue and thus sigWorker won't be woken up unnecessarily. Sounds sensible?

File: conn.go (hunk 17 above)
Comment (krnowak) [inline conn.go]: Argh, forgot about this one. I'll fix it.

File: conn.go (hunk 18 above)
Comment (krnowak) [inline conn.go]: To avoid duplicated channels in conn.signals.

File: conn.go (hunk 15 above)
Comment (bcwaldon) [inline conn.go]: It's unnecessary cognitive overhead. Just drop it :)

File: conn.go (hunk 16 above)
Comment (bcwaldon) [inline conn.go]: Ah, there's a race during shutdown between these worker threads and the shared Conn object. Could you add a comment in here to that effect?

File: conn.go (hunk 18 above)
Comment (bcwaldon) [inline conn.go]: I'm sorry, I don't follow. Explain it to me like I'm an idiot.

Comment (krnowak) [inline conn.go]: But to be clear - you are asking about processPendingSignalChannels function in general or just nilling the signalsMap?

If the former then:
You can use conn.Signal to add or remove a signal channel. When you pass a channel here for the first time, then conn owns it (that means - conn is going to close this channel in conn.Close). If you pass it again, conn will release ownership, so channel is yours again. That's what actually Signal's docs are saying. Before my changes, in spite of docs, this function was just blindly adding passed channel to signal channels list. So somebody could call conn.Signal twice with the same channel and expect that conn is not going to send signals there while in reality the channel appeared twice in the signals list. So every signal could be received twice though that channel. And calling the conn.Close always panicked because of closing the closed channel.

So processPendingSignalChannels simply adds new channel to signal channels list if it wasn't there before or removes it otherwise. And I'm doing it by putting all channels into set, adding or removing them from the set and then creating new signal channels list from the set.

You could see TestRemoveSignal in signal_test.go. It is testing this case.

If the latter then well. Probably not needed.

File: conn.go (hunk 19 above)
Comment (krnowak) [inline conn.go]: Now I don't follow. Copy a channel?

But in general assumption is that if you pass a channel to conn.Signal then conn takes it over (it closes it in conn.Close). You later just use it for listening to channels.

File: conn.go (hunk 21 above)
Comment (krnowak) [inline conn.go]: Probably not necessary, but I don't think it hurts.

File: conn.go (hunk 20 above)
Comment (krnowak) [inline conn.go]: I understand that you can treat conn.signalQueueWakeUp as a pointer to channel. So tmp holds a pointer to channel and conn.signalQueueWakeUp is nil. So from this place sigWorker can call select on either nilled conn.signalQueueWakeUp (which is blocking forever) or on closed conn.shutdownSigWorker (which returns immediately).

File: conn.go (hunk 21 above)
Comment (bcwaldon) [inline conn.go]: It hurts in the sense that people are going to ask this question every time they see this line. Let's just remove the code if it's not necessary.

File: conn.go (hunk 20 above)
Comment (bcwaldon) [inline conn.go]: Ok, that makes sense. And from my local testing, it appears safe to set a channel to nil when something is already `select`ing on it.

File: conn.go (hunk 19 above)
Comment (bcwaldon) [inline conn.go]: Side note: the caller should typically (imagine me waving my hands around) end up closing the channel when she no longer wants to use it rather than allowing another process to close it. It may not matter in this case, though.

File: conn.go (hunk 18 above)
Comment (bcwaldon) [inline conn.go]: Thanks.

**Comment by bcwaldon** [comment]
@krnowak Looks like we're just about done here. Thanks for the hand-holding - this codebase isn't exactly straightforward.

File: conn.go (hunk 19 above)
Comment (krnowak) [inline conn.go]: You can close the channel only after you take it back from connection. Otherwise if you close it, then connection will try to send a signal to it, which results in panic. The same is with eavesdropping API. Before closing the channel used for eavesdropping, you have to call conn.Eavesdrop(nil).

I suppose it would be fixed by introducing an API like:
```go
func signalHandler(signal *dbus.Signal) {
...
}
conn := getConn()
listener := conn.signalListener(signalHandler)
// later, when we don't want to listen to signals anymore
listener.Close()
```

Or just be careful.

Comment (bcwaldon) [inline conn.go]: My original comment was getting at the issue of not knowing how channels act when used in a `range`. To illustrate my example, consider [0] compared to [1]. In example [0], the resulting slice ends up with three pointers all pointing to the same memory address. In example [1] (which simulates your code), the resulting object looks like we expect. My suggestion that we might "need to copy `ch`" is illustrated by the commented out line in example [0] where `v` is copied before a pointer is created and appended to the slice.

So, it took all this to say LGTM :)

[0] http://play.golang.org/p/iV0lurQvgq
[1] http://play.golang.org/p/0jlrm7ZMUk

**Comment by blixtra** [comment]
Can this be merged or is there more clean up that needs to be done?

**Comment by bcwaldon** [comment]
@blixtra @krnowak Looks good to me. @jonboulle or @philips ?

**Comment by philips** [comment]
cc @alexlarsson too

**Comment by philips** [comment]
lgtm, it is sort of a mess but that is no fault of this patch.

**Comment by alexlarsson** [comment]
I'm not super fond of this approach. DBus gives very nice ordering guarantees, and dispatching the signals from a different thread makes it quite likely that things get reordered. On the other hand, since we're using channels here anyway we could already be causing reordering...

Wouldn't it be possible to just make a copy of the signal channels array in the lock and then send to the channels outside the lock. That should make the deadlocking go away, no?

**Comment by krnowak** [comment]
> Wouldn't it be possible to just make a copy of the signal channels array in the lock and then send to the channels outside the lock. That should make the deadlocking go away, no?

Hm, not sure if that would work... Sounds racy. Thread A creates a copy of signal channels array, then thread B removes a signal channel from the original array and closes it, then thread A tries to send stuff to channels inside the copied array, where one channel is closed. That results in panic. Maybe if we sprinkled liberally some locks here and there then maybe something would work, dunno. I haven't been thinking about it too hard right now.

Other solution would be just what we do with eavesdropping channel - if sending to channel would block, then we don't send anything. I decided against doing that.

**Comment by krnowak** [comment]
Hmm, annoyingly the same race exists in my version too. Thread A applies changes to signal channels array in processPendingSignalChannels, then thread B calls Signal to remove channel from the array and closes it, then thread A tries to send stuff in handleSignal to an already closed channel. Bah.

I suppose that a solution like in https://github.com/godbus/dbus/pull/12#discussion_r23889259 would help.

**Comment by kovetskiy** [comment]
any progress in this issue?

**Comment by jsouthworth** [comment]
I'm closing this PR due to age and inactivity. If you would like to still see the changes and they would still be useful please rebase them and open a new PR for discussion.
